src/app/clusters/bindings/bindings.cpp

```diff
@@ -20,155 +20,265 @@
  * @brief Implementation for the Binding Server Cluster
  ***************************************************************************/
 
-#include <app/util/af.h>
-
 #include <app-common/zap-generated/cluster-objects.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/AttributeAccessInterface.h>
 #include <app/CommandHandler.h>
-#include <app/ConcreteCommandPath.h>
+#include <app/ConcreteAttributePath.h>
 #include <app/clusters/bindings/BindingManager.h>
+#include <app/util/attribute-storage.h>
 #include <app/util/binding-table.h>
 #include <lib/support/logging/CHIPLogging.h>
 
 using namespace chip;
-using namespace chip::app::Clusters::Binding;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using chip::app::DataModel::Nullable;
+using TargetStructType         = Binding::Structs::TargetStruct::Type;
+using DecodableBindingListType = Binding::Attributes::BindingList::TypeInfo::DecodableType;
 
 // TODO: add binding table to the persistent storage
+namespace {
 
-static EmberStatus getBindingIndex(EmberBindingTableEntry & newEntry, uint8_t * bindingIndex)
+class BindingTableAccess : public AttributeAccessInterface
 {
-    EmberBindingTableEntry currentEntry;
-    for (uint8_t i = 0; i < EMBER_BINDING_TABLE_SIZE; i++)
+public:
+    // Register for the User Label cluster on all endpoints.
+    BindingTableAccess() : AttributeAccessInterface(Optional<EndpointId>::Missing(), Binding::Id) {}
+
+    CHIP_ERROR Read(const ConcreteReadAttributePath & path, AttributeValueEncoder & encoder) override;
+    CHIP_ERROR Write(const ConcreteDataAttributePath & path, AttributeValueDecoder & decoder) override;
+
+private:
+    CHIP_ERROR ReadBindingTable(EndpointId endpoint, AttributeValueEncoder & encoder);
+    CHIP_ERROR WriteBindingTable(const ConcreteDataAttributePath & path, AttributeValueDecoder & decoder);
+};
+
+BindingTableAccess gAttrAccess;
+
+bool IsValidBinding(const TargetStructType & entry)
+{
+    return (!entry.groupId.HasValue() && entry.endpointId.HasValue() && entry.nodeId.HasValue()) ||
+        (!entry.endpointId.HasValue() && !entry.nodeId.HasValue() && entry.groupId.HasValue());
+}
+
+CHIP_ERROR CheckValidBindingList(const DecodableBindingListType & bindingList)
+{
+    size_t listSize = 0;
+    auto iter       = bindingList.begin();
+    while (iter.Next())
     {
-        emberGetBinding(i, &currentEntry);
-        if (currentEntry.type != EMBER_UNUSED_BINDING && currentEntry == newEntry)
+        VerifyOrReturnError(IsValidBinding(iter.GetValue()), CHIP_ERROR_INVALID_ARGUMENT);
+        listSize++;
+    }
+    ReturnErrorOnFailure(iter.GetStatus());
+    ReturnErrorCodeIf(listSize > EMBER_BINDING_TABLE_SIZE, CHIP_ERROR_NO_MEMORY);
```

Comment:
This check is not enough, right?  We need to compare listSize to the number of list entries not taken up by other-fabric bindings, right?

Also, per spec the failure here is RESOURCE_EXHAUSTED, so that would be `CHIP_IM_GLOBAL_STATUS(ResourceExhausted)`.